natlas-server/app/host/routes.py

```diff
@@ -0,0 +1,218 @@
+@bp.route('/<ip>/screenshots/')
+@isAuthenticated
+def host_screenshots(ip):
+	page = int(request.args.get('p', 1))
+	searchOffset = current_user.results_per_page * (page - 1)
+
+	delHostForm = DeleteForm()
+	rescanForm = RescanForm()
+	info, context = hostinfo(ip)
+	total_entries, screenshots = current_app.elastic.get_host_screenshots(ip, current_user.results_per_page, searchOffset)
+
+	next_url = url_for('host.host_screenshots', ip=ip, p=page + 1) \
+		if total_entries > page * current_user.results_per_page else None
+	prev_url = url_for('host.host_screenshots', ip=ip, p=page - 1) \
+		if page > 1 else None
+
+	version = determine_data_version(context)
+	template_str = f"host/versions/{version}/screenshots.html"
+	return render_template(
+		template_str,
+		**context,
+		historical_screenshots=screenshots,
+		numresults=total_entries,
+		info=info,
+		delHostForm=delHostForm,
+		rescanForm=rescanForm,
+		next_url=next_url,
+		prev_url=prev_url
+	)
+
+
+@bp.route('/<ip>/rescan', methods=['POST'])
+# login_required ensures that an actual user is logged in to make the request
+# opposed to isAuthenticated checking site config to see if login is required first
+@login_required
+def rescan_host(ip):
+	rescanForm = RescanForm()
+
+	if not rescanForm.validate_on_submit():
+		flash("Form failed to validate", "danger")
+		return redirect(request.referrer)
+
+	if not current_app.ScopeManager.isAcceptableTarget(ip):
+		# Someone is requesting we scan an ip that isn't allowed
+		flash(f"We're not allowed to scan {ip}", "danger")
+		return redirect(request.referrer)
+
+	incompleteScans = current_app.ScopeManager.getIncompleteScans()
+
+	scan_dispatched = {}
+	for scan in incompleteScans:
+		scan_dispatched[scan.target] = scan.dispatched
+
+	if ip in scan_dispatched:
+		if scan_dispatched[ip]:
+			status = "dispatched"
+			if (datetime.utcnow() - scan.date_dispatched).seconds > 1200:
+				# 20 minutes have past since dispatch, something probably went wrong
+				# move it back to not dispatched and update the cached rescan data
+				scan.dispatched = False
+				db.session.add(scan)
+				db.session.commit()
+				current_app.ScopeManager.updatePendingRescans()
+				current_app.ScopeManager.updateDispatchedRescans()
+				flash(f"Refreshed existing rescan request for {ip}", "success")
+				return redirect(request.referrer)
+		else:
+			status = "pending"
+		flash(f"There's already a {status} rescan request for {ip}", "warning")
+		return redirect(request.referrer)
+
+	rescan = RescanTask(user_id=current_user.id, target=ip)
+	db.session.add(rescan)
+	db.session.commit()
+	current_app.ScopeManager.updatePendingRescans()
+	current_app.ScopeManager.updateDispatchedRescans()
+	flash(f"Requested rescan of {ip}", "success")
+	return redirect(request.referrer)
+
+
+@bp.route("/random")
+@bp.route("/random/")
+def random_host():
+	random_host = current_app.elastic.random_host()
+	if not random_host:
+		abort(404)
+	ip = random_host['hits']['hits'][0]['_source']['ip']
```

Comment:
random_host() should directly return ['hits']['hits'] or ['hits']['hits'][0]

Right now random_host() is a leaky abstraction and is exposing the internal Elasticsearch response type, but it should up the abstraction a bit and return a rich object.

I bet this API is going to overlap a lot with the `/<ip>`. That suggests to me that we should either 301 Redirect to the `/<ip>` endpoint or extract this out into a shared method.

Reply:
You're right, this is going to be pretty identical to the `/<ip>` route, but I avoided doing a redirect because I wanted the url to be able to stay `/random` and be able to hit f5 to get a new host each refresh, instead of having to click random at the top each time. I think we can definitely abstract it out a bit and improve the template rendering at the same time, but let's save that for another time because right now this PR touches a ton of files and I don't think we should drag it on for too long and create a bunch of merge conflicts for other work.